# 원본 문제

In [ ]:
"""## DNN(MLP) 모델을 이용한 보스터 집값 회귀 분석
* 체점 기준 :
  - 데이터 셋 : 체점 서버 내 테스트 데이터 셋
  - 성능 지표 : MAE
  - PASS 기준 : 3.0 이하
  """

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

from tensorflow.keras.datasets import boston_housing

"""* Step 1. Inptu tensor 와 Target tensor 준비(훈련데이터)"""

# 수강생 작성 코드
# 1. import 한 boston_housing API를 이용하여 boston_housing 데이터 셋을 다운로드 받으세요)


"""* Step 2. 데이터의 전처리 """

# 수강생 작성 코드
# 1. train 데이터의 feature 별 평균값, 표준편차 값을 이용하여 정규화 작업을 진행하세요



"""* Step 3. batch 구성을 위한 데이터 pipeline 생성"""

# 수강생 작성 코드
# 1. 훈련 데이터, 테스트 데이터를 파이토치의 텐서 형태로 변환해주세요


# 수강생 작성 코드
# 1. torch.utils.data.TensorDataset API를 활용해 훈련, 검증, 테스트 데이터를 위한 Dataset 객체를 생성 하세요
# 2. 훈련 Dataset을 torch.utils.data.random_split() API 를 이용해 5:1로 분할하여 검증 Dataset객체를 생성 하세요


# 수강생 작성 코드
# 1. torch.utils.data.DataLoader API를 활용해 훈련, 검증, 테스트 데이터 pipeline을 생성하세요



"""* Step 4. DNN(MLP) 모델 디자인"""

# 수강생 작성 코드
# 1. torch.nn.Module API를 이용하여 보스턴 집값 회귀 분석 데이터 셋을 분석 하기 위한 MLP 모델 class를 정의 하세요
# 2. 정의한 MLP 모델 class를 이용하여 모델 객체를 생성 하세요



"""* Step 5. 모델의 학습 정보 설정"""

# 수강생 작성 코드
# 1. torch.nn 모듈에 구현된 API를 이용해 회귀를 위한 손실함수 객체를 생성하세요
# 2. torch.nn 모듈에 구현된 API를 이용해 체점 기준 MAE 를 계산하기 위한 객체를 생성하세요
# 3. torch.optim API로 옵티마이저를 설정하세요


"""* Step 6. 모델의 학습 loop 를 정의하는 함수 생성"""

def train(epoch):
  # 수강생 작성 코드
  # 1. 모델 객체를 학습에 적합한 상태로 변경하세요


  # 수강생 작성 코드
  # 1. 학습데이터 에서 미니 배치를 추출하여 학습 loop를 수행하는 for문을 작성 하세요
  for batch_idx,(____, ____) in enumerate(____):
    # 수강생 작성 코드
    # 1. 모델에 input data를 전달하여 순전파를 수행하세요


    # 수강생 작성 코드
    # 1. loss 함수에 모델의 예측값 과 정답 정보를 전달하여 loss 값을 계산 하세요


    # 수강생 작성 코드
    # 1. mae 계산 함수에 모델의 예측값 과 정답 정보를 전달하여 회귀 분석의 성능지표 MAE 계산 하세요


    # 수강생 작성 코드
    # 1. optimizer 객체를 이용해 모델을 구성하는 파라미터들(w, b)의 gradient를 초기화 하세요


    # 수강생 작성 코드
    # 1. 모델을 구성하는 파라미터들(w, b)이 loss 에 미치는 영향도(gradient)를 계산 하세요


    # 수강생 작성 코드
    # 1. 계산된 영향도(gradient) 값을 이용하여 모델의 파라미터(w,b)를 업데이트 하세요


    # 모델의 성능 지표를 계산 및 출력 하여 학습 진행을 체크하는 코드
    if batch_idx%50 == 0:
      print('Train Epoch: {} [{}/{} ({:.0f}%)]\tMAE: {:.6f}\tMSE: {:.06f}'.format(
            epoch,
            (batch_idx+1) * len(data),
            len(train_loader.dataset),
            100. * (batch_idx+1) / len(train_loader),
            batch_mae,
            batch_mse
            ))


"""* Step 7. 모델의 검증 loop 를 정의하는 함수 생성"""

def test(data_loader):
  # 수강생 작성 코드
  # 1. 모델 객체를 테스트에 적합한 상태로 변경하세요


  # 수강생 작성 코드
  # 1. 모델의 성능을 지표를 계산에 필요한 정보 수집을 위한 변수 선언
  #   - loss 를 수집하기 위한 'test_mse' 변수 선언 및 0 으로 초기화
  #   - MAE 를 수집하기 위한 'test_mae' 변수 선언 및 0 으로 초기화


  # 수강생 작성 코드
  # 1. 검증데이터 에서 미니 배치를 추출하여 검증 loop를 수행하는 for문을 작성 하세요
  for ____, ____ in ____:
    # 수강생 작성 코드
    # 1. 모델에 input data를 전달하여 순전파를 수행하세요


    # 수강생 작성 코드
    # 1. loss 함수에 모델의 예측값 과 정답 정보를 전달하여 모델의 성능을 검증하기 위한 loss 값을 계산 하세요
    #   - 배치별로 계산된 loss 정보를 'test_mse' 변수에 누적
    # 2. MAE 함수에 모델의 예측값 과 정답 정보를 전달하여 모델의 성능을 검증하기 위한 MAE 값을 계산 하세요
    #   - 배치별로 계산된 MAE 정보를 'test_mae' 변수에 누적


  # 수강생 작성 코드
  # 1. 'test_mae' 에 누적된 배치 별 MAE의 합을 배치의 개수로 나누에 전체 데이터 셋의 평균 MAE를 계산 하세요
  #    - 계산 된 평균 MAE를 'test_mae' 변수에 할당
  # 2. 'test_mse' 에 누적된 배치 별 MAE의 합을 배치의 개수로 나누에 전체 데이터 셋의 평균 loss를 계산 하세요
  #    - 계산 된 평균 MAE를 'test_mse' 변수에 할당

  # 모델의 성능 지표를 계산 및 출력 하여 학습 진행을 체크하는 코드
  print('\nTest set: Average MAE: {:.4f}\t MSE: {:.4f}\n'.format(
        test_mae,
        test_mse
        ))


"""* Step 5. 훈련 함수와 검증 함수로 모델 훈련 및 검증"""
# 수강생 작성 코드
# 1. 반복문을 활용해 epoch 수 만큼 훈련 및 검증을 하세요.


# 수강생 작성 코드
# 1. 테스트 데이터 셋을 이용해 모델의 성능을 확인 하세요


# 풀이

In [ ]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

from tensorflow.keras.datasets import boston_housing

In [ ]:
"""* Step 1. Inptu tensor 와 Target tensor 준비(훈련데이터)"""

# 수강생 작성 코드
# 1. import 한 boston_housing API를 이용하여 boston_housing 데이터 셋을 다운로드 받으세요
(train_x, train_y), (test_x, test_y) = boston_housing.load_data()

57026/57026 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [ ]:
"""* Step 2. 데이터의 전처리 """

# 수강생 작성 코드
# 1. train 데이터의 feature 별 평균값, 표준편차 값을 이용하여 정규화 작업을 진행하세요
# test, train 독립변수는 normalization을 해야함 but 종속변수는 할 필요x
# 평균값 및 표준편차 값 구하기
x_mean, x_std = train_x.mean(axis=0), train_x.std(axis=0)

# 정규화 작업
train_x = (train_x - x_mean)/x_std

test_x = (test_x - x_mean)/x_std

In [ ]:
"""* Step 3. batch 구성을 위한 데이터 pipeline 생성"""

# 수강생 작성 코드
# 1. 훈련 데이터, 테스트 데이터를 파이토치의 텐서 형태로 변환해주세요
# 회귀 문제에서는 값을 예측하는 것 -> float 대신에 long
# nn.Linear가 2D tensor을 요구하기 때문에 (batch_size, output_dim) 차원을 1D -> 2D
train_x, train_y = torch.Tensor(train_x), torch.Tensor(train_y).view(-1,1)
test_x, test_y = torch.Tensor(test_x), torch.Tensor(test_y).view(-1,1)

# 수강생 작성 코드
# 1. torch.utils.data.TensorDataset API를 활용해 훈련, 검증, 테스트 데이터를 위한 Dataset 객체를 생성 하세요
# 2. 훈련 Dataset을 torch.utils.data.random_split() API 를 이용해 5:1로 분할하여 검증 Dataset객체를 생성 하세요
train_dataset = torch.utils.data.TensorDataset(train_x, train_y )
train_dataset_count = int(len(train_dataset)*(5/6))
val_dataset_count = len(train_dataset) - train_dataset_count

train_dataset, val_dataset = torch.utils.data.random_split(train_dataset, [train_dataset_count, val_dataset_count])
test_dataset = torch.utils.data.TensorDataset(test_x, test_y)

# 수강생 작성 코드
# 1. torch.utils.data.DataLoader API를 활용해 훈련, 검증, 테스트 데이터 pipeline을 생성하세요
batch_size = 1

train_loader = torch.utils.data.DataLoader(train_dataset,
                                           batch_size=batch_size,
                                           shuffle=True)
val_loader = torch.utils.data.DataLoader(val_dataset,
                                         batch_size=batch_size,
                                         shuffle=False)
test_loader = torch.utils.data.DataLoader(test_dataset,
                                         batch_size=batch_size,
                                         shuffle=False)

In [ ]:
train_x.shape, test_x.shape, train_y.shape, test_y.shape

(torch.Size([404, 13]),
 torch.Size([102, 13]),
 torch.Size([404, 1]),
 torch.Size([102, 1]))

In [ ]:
x, y =next(iter(train_loader))

In [ ]:
x.size()

torch.Size([8, 13])

In [ ]:
"""* Step 4. DNN(MLP) 모델 디자인"""

# 수강생 작성 코드
# 1. torch.nn.Module API를 이용하여 보스턴 집값 회귀 분석 데이터 셋을 분석 하기 위한 MLP 모델 class를 정의 하세요
# 2. 정의한 MLP 모델 class를 이용하여 모델 객체를 생성 하세요
class Model_boston(nn.Module):
  def __init__(self):
    super().__init__()
    self.fc1 = nn.Linear(13, 64)
    self.fc2 = nn.Linear(64, 32)
    self.fc3 = nn.Linear(32, 1)

  def forward(self, x):
      x = F.relu(self.fc1(x))
      x = F.relu(self.fc2(x))
      x = self.fc3(x)
      return x

In [ ]:
# 수강생 작성 코드
# 1. torch.nn 모듈에 구현된 API를 이용해 회귀를 위한 손실함수 객체를 생성하세요
# 2. torch.nn 모듈에 구현된 API를 이용해 체점 기준 MAE 를 계산하기 위한 객체를 생성하세요
# 3. torch.optim API로 옵티마이저를 설정하세요

model = Model_boston()
criterion_mse = nn.MSELoss()
criterion_mae = nn.L1Loss()
optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)


In [ ]:
"""* Step 6. 모델의 학습 loop 를 정의하는 함수 생성"""
def train(epoch):
  # 수강생 작성 코드
  # 1. 모델 객체를 학습에 적합한 상태로 변경하세요
  model.train()

  batch_mse, batch_mae = 0, 0
  # 수강생 작성 코드
  # 1. 학습데이터 에서 미니 배치를 추출하여 학습 loop를 수행하는 for문을 작성 하세요
  for batch_idx,(data, target) in enumerate(train_loader):
    # 수강생 작성 코드
    # 1. 모델에 input data를 전달하여 순전파를 수행하세요
    # model.forward를 할 필요x : nn.module이 __call__() method를 통해 foward를 자동으로 불러줌
    output = model(data)

    # 수강생 작성 코드
    # 1. loss 함수에 모델의 예측값 과 정답 정보를 전달하여 loss 값을 계산 하세요
    batch_mse = criterion_mse(output, target)

    # 수강생 작성 코드
    # 1. mae 계산 함수에 모델의 예측값 과 정답 정보를 전달하여 회귀 분석의 성능지표 MAE 계산 하세요
    batch_mae = criterion_mae(output, target)

    # 수강생 작성 코드
    # 1. optimizer 객체를 이용해 모델을 구성하는 파라미터들(w, b)의 gradient를 초기화 하세요
    optimizer.zero_grad()

    # 수강생 작성 코드
    # 1. 모델을 구성하는 파라미터들(w, b)이 loss 에 미치는 영향도(gradient)를 계산 하세요
    batch_mse.backward()

    # 수강생 작성 코드
    # 1. 계산된 영향도(gradient) 값을 이용하여 모델의 파라미터(w,b)를 업데이트 하세요
    optimizer.step()

    # 모델의 성능 지표를 계산 및 출력 하여 학습 진행을 체크하는 코드
    if batch_idx%50 == 0:
      print('Train Epoch: {} [{}/{} ({:.0f}%)]\tMAE: {:.6f}\tMSE: {:.06f}'.format(
            epoch,
            (batch_idx+1) * len(data),
            len(train_loader.dataset),
            100. * (batch_idx+1) / len(train_loader),
            batch_mae,
            batch_mse
            ))


In [ ]:
def test(data_loader):
  # 수강생 작성 코드
  # 1. 모델 객체를 테스트에 적합한 상태로 변경하세요
  model.eval()

  # 수강생 작성 코드
  # 1. 모델의 성능을 지표를 계산에 필요한 정보 수집을 위한 변수 선언
  #   - loss 를 수집하기 위한 'test_mse' 변수 선언 및 0 으로 초기화
  #   - MAE 를 수집하기 위한 'test_mae' 변수 선언 및 0 으로 초기화
  test_mse =0
  test_mae = 0

  # 수강생 작성 코드
  # 1. 검증데이터 에서 미니 배치를 추출하여 검증 loop를 수행하는 for문을 작성 하세요
  for data, target in data_loader:
    # 수강생 작성 코드
    # 1. 모델에 input data를 전달하여 순전파를 수행하세요
    output = model(data)

    # 수강생 작성 코드
    # 1. loss 함수에 모델의 예측값 과 정답 정보를 전달하여 모델의 성능을 검증하기 위한 loss 값을 계산 하세요
    #   - 배치별로 계산된 loss 정보를 'test_mse' 변수에 누적
    # 2. MAE 함수에 모델의 예측값 과 정답 정보를 전달하여 모델의 성능을 검증하기 위한 MAE 값을 계산 하세요
    #   - 배치별로 계산된 MAE 정보를 'test_mae' 변수에 누적
    test_mse += criterion_mse(output, target).item()
    test_mae += criterion_mae(output, target).item()

  # 수강생 작성 코드
  # 1. 'test_mae' 에 누적된 배치 별 MAE의 합을 배치의 개수로 나누에 전체 데이터 셋의 평균 MAE를 계산 하세요
  #    - 계산 된 평균 MAE를 'test_mae' 변수에 할당
  # 2. 'test_mse' 에 누적된 배치 별 MAE의 합을 배치의 개수로 나누에 전체 데이터 셋의 평균 loss를 계산 하세요
  #    - 계산 된 평균 MAE를 'test_mse' 변수에 할당
  test_mae /= len(data_loader)
  test_mse /= len(data_loader)

  # 모델의 성능 지표를 계산 및 출력 하여 학습 진행을 체크하는 코드
  print('\nTest set: Average MAE: {:.4f}\t MSE: {:.4f}\n'.format(
        test_mae,
        test_mse
        ))

In [ ]:
"""* Step 7. 훈련 함수와 검증 함수로 모델 훈련 및 검증"""
# 수강생 작성 코드
# 1. 반복문을 활용해 epoch 수 만큼 훈련 및 검증을 하세요.
epochs = 600
for epoch in range(1,epochs+1):
  train(epoch)
  test(val_loader)

# 수강생 작성 코드
# 1. 테스트 데이터 셋을 이용해 모델의 성능을 확인 하세요
test(test_loader)

Streaming output truncated to the last 5000 lines.
Train Epoch: 101 [151/336 (45%)]	MAE: 2.476322	MSE: 6.132172
Train Epoch: 101 [201/336 (60%)]	MAE: 3.322662	MSE: 11.040085
Train Epoch: 101 [251/336 (75%)]	MAE: 1.194485	MSE: 1.426794
Train Epoch: 101 [301/336 (90%)]	MAE: 0.336384	MSE: 0.113154

Test set: Average MAE: 2.6373	 MSE: 17.5796

Train Epoch: 102 [1/336 (0%)]	MAE: 4.124217	MSE: 17.009167
Train Epoch: 102 [51/336 (15%)]	MAE: 0.661219	MSE: 0.437210
Train Epoch: 102 [101/336 (30%)]	MAE: 6.461857	MSE: 41.755592
Train Epoch: 102 [151/336 (45%)]	MAE: 1.060199	MSE: 1.124021
Train Epoch: 102 [201/336 (60%)]	MAE: 1.698318	MSE: 2.884286
Train Epoch: 102 [251/336 (75%)]	MAE: 1.862110	MSE: 3.467454
Train Epoch: 102 [301/336 (90%)]	MAE: 1.278259	MSE: 1.633947

Test set: Average MAE: 2.6247	 MSE: 17.5296

Train Epoch: 103 [1/336 (0%)]	MAE: 1.384243	MSE: 1.916129
Train Epoch: 103 [51/336 (15%)]	MAE: 0.574741	MSE: 0.330328
Train Epoch: 103 [101/336 (30%)]	MAE: 0.022556	MSE: 0.000509
Train Ep

----------
20250310
- 회귀 모델 제작 과정을 다 구현했다.
- hyperparameter
  - batch size는 data 수가 적어서 8, 16이 제일 잘 나온다
  - learning rate 1e-3, 1e-4가 제일 좋게 나오는 듯하다

- model structure
  - 이제 내일 하자..

- 총정리
  - 3/10 ~ 3/16 내로 하자

-----
20250313
- HYPERPARMETER
  - 데이터의 수가 336밖에 없어서 그냥 전체로 학습시킴 -> BATCH_SIZE = 1
  - EPOCH를 100 ~ 200 정도하다가 중도 LOSS가 순회하는 것 같아 아예 큰 값인 1000으로 실행 -> 괜춘?
  
- 결론 : STRUCTURE을 안 바꿔도 HYPERPARAMETER만 바꿔도 목표점수를 넘을 수 있었다